### ISSUE TITLE
EdDSA Signature only verifies first 64 bytes

### ISSUE THREAD

**Issue by mpbarnwell** [bcgit/bc-java#508]
When verifying signatures with EdDSA (I've only tested with Ed25519), the verification process only looks at the first 64 bytes of the signature, so if I pass in longer signatures, they will pass verification - this feels wrong to me, although I don't believe it's a vulnerability. 

I've included a test below using the example found in [RFC8032 Test 2](https://tools.ietf.org/html/rfc8032#page-24), in particular:

Public key: `3d4017c3e843895a92b70aa74d1b7ebc9c982ccf2ec4968cc0cd55f12af4660c`
Message: `72`
Signature:  `92a009a9f0d4cab8720e820b5f642540a2b27b5416503f8fb3762223ebdb69da085ac1e43e15996e458f3613d0f11d8c387b2eaeb4302aeeb00d291612bb0c00`

The signature correctly validates but so does the same signature with random bytes appended, e.g.

Garbage Signature:  `92a009a9f0d4cab8720e820b5f642540a2b27b5416503f8fb3762223ebdb69da085ac1e43e15996e458f3613d0f11d8c387b2eaeb4302aeeb00d291612bb0c00abba`

I would suggest a length check on the input to `implVerify` in https://github.com/bcgit/bc-java/blob/master/core/src/main/java/org/bouncycastle/math/ec/rfc8032/Ed25519.java#L400

```java
import com.google.common.io.BaseEncoding;
import org.bouncycastle.asn1.edec.EdECObjectIdentifiers;
import org.bouncycastle.asn1.x509.AlgorithmIdentifier;
import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.junit.Test;

import java.io.IOException;
import java.security.*;
import java.security.spec.X509EncodedKeySpec;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class SigTest {

    @Test
    public void testSignature() throws GeneralSecurityException, IOException {
        BouncyCastleProvider bouncyCastleProvider = new BouncyCastleProvider();
        Security.addProvider(bouncyCastleProvider);

        // Key taken from TEST 2 of RFC8032 - https://tools.ietf.org/html/rfc8032#page-24
        String hexEncodedPubKey = "3d4017c3e843895a92b70aa74d1b7ebc9c982ccf2ec4968cc0cd55f12af4660c";

        // Convert to JCA format
        byte[] pubKeyBytes = BaseEncoding.base16().lowerCase().decode(hexEncodedPubKey);
        SubjectPublicKeyInfo pubKeyInfo = new SubjectPublicKeyInfo(
                new AlgorithmIdentifier(EdECObjectIdentifiers.id_Ed25519), pubKeyBytes);
        X509EncodedKeySpec keySpec = new X509EncodedKeySpec(pubKeyInfo.getEncoded());
        KeyFactory keyFactory = KeyFactory.getInstance("Ed25519", bouncyCastleProvider);
        PublicKey pk = keyFactory.generatePublic(keySpec);

        // Sign test data
        byte[] expectedSignature = BaseEncoding.base16().lowerCase().decode(
                "92a009a9f0d4cab8720e820b5f642540" +
                        "a2b27b5416503f8fb3762223ebdb69da" +
                        "085ac1e43e15996e458f3613d0f11d8c" +
                        "387b2eaeb4302aeeb00d291612bb0c00");
        byte[] data = { 0x72 };

        Signature signature = Signature.getInstance("Ed25519", bouncyCastleProvider);
        signature.initVerify(pk);
        signature.update(data);
        assertTrue(signature.verify(expectedSignature));

        // Add garbage to end of expected signature
        byte[] garbage = { 0xA, 0xB, 0xB, 0xA };
        byte[] garbageSignature = new byte[expectedSignature.length + garbage.length];
        System.arraycopy(expectedSignature, 0, garbageSignature, 0, expectedSignature.length);
        System.arraycopy(garbage, 0, garbageSignature, expectedSignature.length, garbage.length);

        // Re-init signature
        signature = Signature.getInstance("Ed25519", bouncyCastleProvider);
        signature.initVerify(pk);
        signature.update(data);
        assertFalse(signature.verify(garbageSignature)); // Expect to be false, but returns true
    }

}
```

**Comment by peterdettman** [comment]
Thanks for the report. A fix has been pushed to our git and should mirror here soon.

**Comment by peterdettman** [comment]
Fixed in 1.62
